### ISSUE TITLE
Uncaught exception when CompletableDeferred is completedExceptionally

### ISSUE THREAD

**Issue by RegalSkeppet** [Kotlin/kotlinx.coroutines#251]
I have the following test where I expect that all lines printed should start with "expected:" and that the CoroutineExceptionHandler is never invoked. But this is not what actually happens. I'm assuming this is a race when cancelling with timeout and completingExceptionally happens at the same time. The documentation says CompleteableDeferred is thread-safe. Or am I missing something and this is the correct behavior?

```
import kotlinx.coroutines.experimental.*
import kotlin.test.Test
import kotlin.test.assertNull

class CoroutineTests {
    @Test
    fun test() = runBlocking<Unit>(CoroutineExceptionHandler { c, t ->
        println("NOT EXPECTED: $t")
    }) {
        val delayMillis = 10
        repeat(100) {
            val d = CompletableDeferred<Int>()
            repeat(8) { i ->
                launch {
                    delay(delayMillis)
                    d.completeExceptionally(Exception("Exception $i"))
                }
            }
            val r = try {
                withTimeout(delayMillis) {
                    d.await()
                }
            } catch (t: Throwable) {
                println("expected: $t")
                null
            }
            assertNull(r)
        }
    }
}
```

**Comment by elizarov** [comment]
There is a race in your `withTimeout(delayMillis) { d.await() }` coroutine. The race is between the following two events:

* This coroutine can get cancelled on timeout.
* This coroutine can crash when `d.await()` throws exception.

The way that `CoroutineExceptionHandler` currently works is that it is invoked on any exception that the coroutines core don't know how process. It is basically a gimmick to make sure that exceptions are not "lost".

What happens here is that sometimes we get to a state when this `withTimeout` coroutine gets cancelled (on timeout) when it is already in the process of dispatching an exception from `d.await()`. Now, when this exception gets handled, it is supposed to terminate `withTimeout` coroutine, but, by that time, the coroutine is already cancelled (on timeout) and thus `CoroutineExceptionHandler` is invoked to report that we have an exception that we are about to loose here. 

See the code here: https://github.com/Kotlin/kotlinx.coroutines/blob/master/core/kotlinx-coroutines-core/src/main/kotlin/kotlinx/coroutines/experimental/AbstractContinuation.kt#L103

This is really a questionable piece of code, but I frankly have not figured out a good alternative. One approach would be to abandon cancellation cause and instead make coroutine completed with the actual exception. This seems Ok if coroutine is cancelled "normally" (without an explicit cause, on timeout or with some other cause that is an instance of `CancellationException`). 

But what if coroutine was cancelled with a truly important and fatal cause. What if, for example, we have the following sequence of events:
1. `Job.cancel(IOException(...))` was invoked on a coroutine because, for example, a child of this coroutine (that was performing a part of this coroutine's work) has crashed with `IOException`.
2. The coroutine had started to wind down, but had crashed by itself with some other exception.

What should be the ultimate termination cause for this coroutine and how do we make sure neither of exceptions are lost?

**Comment by RegalSkeppet** [comment]
My assumption was that either the timeout or the result of the CompletableDeferred would be used depending on which one happens first. I see that I didn't fully understand how things work, thank you for the explanation. I think I need to read up on the inner workings of coroutines before I can suggest another solution.
So if the code would be changed to use a select-clause with onAwait and onTimeout instead, would it behave as I initially expected?

**Comment by elizarov** [comment]
The result of "whichever happens first" is the rule that is implemented now and it is racy. If you first cancel your coroutine and then it crashed (while cancelling), then the crash gets delivered to the uncaught exception handler. It is currently an inherent problem in coroutines.

**Comment by kondouraion** [comment]
```

import kotlinx.coroutines.experimental.*
import org.junit.Test
import kotlin.test.assertNull

class CoroutineTests {
    @Test
    fun test() = runBlocking<Unit>(CoroutineExceptionHandler { c, t ->
        println("NOT EXPECTED: $t")
    }) {
        val delayMillis = 10
        repeat(100) {
            val d = CompletableDeferred<Int>()
            repeat(8) { i ->
                launch {
                    delay(delayMillis)
                    d.completeExceptionally(Exception("Exception $i"))
                }
            }
            val r = try {
                async(CoroutineExceptionHandler { coroutineContext, throwable ->//new
                    println("Catch: $throwable")//new
                }) {//new
                    withTimeout(delayMillis) {
                        d.await()
                    }
                }.await()//new
            } catch (t: Throwable) {
                println("expected: $t")
                null
            }
            assertNull(r)
        }
    }
}
```

**Comment by kinnrot** [comment]
I guess that when a timeout reached, no matter what happens inside everything must stop, best case, the developer will be able to know what exception occurred on cancellation if any and possible.

Maybe while cancelling a coroutine due to timeout, the cancellation process should swallow all exceptions and report them inside the timeout response object?

**Comment by elizarov** [comment]
This issue does not reproduce anymore since Kotlin 1.1.0. It was fixed together by  #896 (see it for explanation). The second exception in race is now swallowed.
